src/dmd/expressionsem.d

```diff
@@ -5784,6 +5784,27 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
               assert(a == b, _d_assert_fail!"=="(a, b));
             }()
             */
+
+            /// Stores the result of an operand expression into a tempory
+            /// if necessary, e.g. if it is a fuction call as in
+            /// https://issues.dlang.org/show_bug.cgi?id=20114
+            auto maybePromoteToTmp(ref Expression op)
+            {
+                if(op.isCallExp)
```

Comment:
This sounds way too limited. There should be a check for side effect, rather than just function call.

Reply:
also only if `-checkaction=context`

Reply:
@Geod24 Good point, but should strongly `pure` function result in a temporary (as the function may be arbitrary complex?)

Reply:
@WalterBright All changes are guarded by 
https://github.com/dlang/dmd/blob/1cdad0070c85fc9548d7957dc11babacb781084c/src/dmd/expressionsem.d#L5778

Reply:
> od point, but should strongly `pure` function result in a temporary (as the function may be arbitrary complex?)

I don't think it matters, because both gives a correct result. And if you are hitting an `assert` failure, performance don't matter as much. I would say whatever is easiest.

Reply:
Yeah, performance issues should not matter when hitting an assert.
My only concern is wether there are functions for which `hasSideEffect` erroneously returns `false`, e.g. when calling external code. But I am not sure if we should even care about that scenario

Reply:
Ping @Geod24